Add a new index for BHA

AmitPhulera opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
Initial work to create a new index for BHA based of case search index.
This might not be functional, have put the code out there to get initial feedback on the approach. 

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [ ] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [ ] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [ ] Risk label is set correctly
- [ ] The set of people pinged as reviewers is appropriate for the level of risk of the change

AmitPhulera committed:
add new index name in es consts file

AmitPhulera committed:
add case_search_bha adapter

AmitPhulera committed:
Merge remote-tracking branch 'origin/master' into ap/new-index-for-bha

AmitPhulera committed:
add bha adapter config to __init__.py

AmitPhulera committed:
update index name in const.py

AmitPhulera committed:
run ./manage.py make_elastic_migration --name add_new_index_for_bha -t=5 -c case_search_bha:case-search-bha-2024-05-10

AmitPhulera committed:
update setting values for bha index

AmitPhulera committed:
override index method for ElasticCaseSearch

AmitPhulera committed:
remove extra words that were somehow added

esoergel commented:
I'm particularly interested to see how this integrates with pillows and other places that write.  We spoke offline about this a bit already, but I was imagining this to live within the standard case search adapter, so nothing else needs to know about writing to multiple places.

The other thing I was thinking about while reviewing this is how this would work with more than one dedicated index.  Like, in my mind it'd be ideal if there'd be a config option in `localsettings` like
```python
DEDICATED_CASE_SEARCH_INDICES = {
    # domain_name: index_name
    'mydomain': 'case-search-mydomain',
    'anotherdomain': 'case-search-anotherdomain',
}
```
And then this could be used to direct writes as appropriate in the adapter.  Otherwise it feels awfully boilerplate-y to have to make a replica adapter for each.  We could also do likewise for writes (rather than the db based approach I floated in https://github.com/dimagi/commcare-hq/pull/34601), though I think we'll want the two independently configurable in any case.

AmitPhulera committed:
add util to get domain from object which can be dict or object

AmitPhulera committed:
add tests for index operation on case-search adapter

AmitPhulera committed:
add bulk method in case_search adapter to multiplex writes to bha index when required

AmitPhulera committed:
add docs to the newly added functions

AmitPhulera committed:
support domain in delete doc method in pillow and add ability to multiplex deletes if required

AmitPhulera committed:
pass domain in all the places where deletions are happening

AmitPhulera committed:
add index_runtime name in migration file

AmitPhulera committed:
if doc is None return None

this will be the case when we have called bulk_delete and passed it ids, bulk_delete calls bulk. We only use it at one place in HQ which is to delete case search data for a entire domain in HQ which I don't think we will do

AmitPhulera committed:
fix failing tests

snopoke commented:
Like @esoergel I was also expecting more of a settings approach rather than a new adapter but its possible that we would need a new adapter anyway because of how they are tied to migrations etc. Maybe we could create them dynamically though?

AmitPhulera committed:
fix migrations.lock file

AmitPhulera committed:
add check for doc is None

AmitPhulera committed:
failing test fix

AmitPhulera commented:
> I'm particularly interested to see how this integrates with pillows and other places that write. We spoke offline about this a bit already, but I was imagining this to live within the standard case search adapter, so nothing else needs to know about writing to multiple places.

Yeah that is correct atleast for the writes. [ab2ba07](https://github.com/dimagi/commcare-hq/pull/34602/commits/ab2ba077a0630099af4f908bb6ba46bac6fe2b38)  https://github.com/dimagi/commcare-hq/pull/34602/commits/7013b4254394cd3cdca5b7702fb4ae121a6d56e0  [7cb3b24](https://github.com/dimagi/commcare-hq/pull/34602/commits/7cb3b2412a7dfe178269579a2066da5be0576eb5) subsequently ovverrides index and bulk operations 

For deletes I ended up modifying pillows [813a664](https://github.com/dimagi/commcare-hq/pull/34602/commits/813a66430a80a6ea109f03151883c2eb405d6be7) https://github.com/dimagi/commcare-hq/pull/34602/commits/5cb87066343702dd11d703e5ce050e24f7fd3147 as we already have domain information here.


> The other thing I was thinking about while reviewing this is how this would work with more than one dedicated index. Like, in my mind it'd be ideal if there'd be a config option in localsettings. 

I have kept a plug https://github.com/dimagi/commcare-hq/pull/34602/commits/ab2ba077a0630099af4f908bb6ba46bac6fe2b38#diff-7ba2ac2cd9c42581b497ab0c7cb5ea77288d3a8d7e8583cfc2c35a4da9e063b6R196-R199 which can essentially return an adapter based on any condition that seem fit. 

I think we should use settings to figure out the adapter and `CaseSearchConfig` to determine if we have to multiplex writes. Two new attributes `multiplex_writes`, `read_from_new_index` can be added to `CaseSearchConfig` which can be modified at runtime as per requirement. 

And regarding having it for multiple dedicated indices, it should be quite possible by the approach but yes it will add the overhead of the boilerplate and I think it should be okay. My idea of these indices is as of SQL tables having their own separate model classes. As per @snopoke's point we can definitely add more tooling to auto-generate the adapter classes automatically but the question again is how often will it be used. 

> This index applies only to one environment, so we shouldn't build and write to it in other settings

I have this on my radar and will implement it part of sepearte PR where our migration framework will support the option of having environment specific indices. 


>  I'm not certain how best to manage that, but one thing that sounds appealing is treating this as a second case search index, in the sense that all migrations applied to the main case search index are also applied to this one by default. 

This is a great point. I have a vague idea of how it could be done without changing things in commcare-cloud. Will have to think more about it.

AmitPhulera committed:
add fn to decide if to multiplex or not

AmitPhulera commented on corehq/ex-submodules/pillowtop/processors/elastic.py:
```diff
@@ -70,20 +70,21 @@ def process_change(self, change):
 
         if change.deleted and change.id:
             doc = change.get_document()
+            domain = doc.get('domain')
```
I am not very sure about this change and I was also not very sure how to add tests for it. Any inputs here would be super helpful.

AmitPhulera replied:
This was the document that is to be indexed in ES, it should always have a domain. But based on your suggestion I have added https://github.com/dimagi/commcare-hq/pull/34673/commits/3417fef5f2b26b545d3bf40837e06c08d93c16fc#diff-308591ed80f437c63d421731473742a6d8fec03d88c9410b46d1d39704dcc66dR73-R75

AmitPhulera committed:
Merge remote-tracking branch 'origin/master' into ap/new-index-for-bha

AmitPhulera commented on corehq/apps/es/case_search.py:
```diff
@@ -179,6 +179,8 @@ def _get_domain_from_doc(self, doc):
         This will fail hard if domain is not present in doc.
         """
+        if not doc:
+            return None
```
I am not actually sure if I can confidently say I am aware of all the paths. But if we delete a domain which is on sub-index then we will have a couple of manual steps to do first i.e stop writes on that index by updating the config and index deletion can be taken care of then rather than automating it?

AmitPhulera commented:
I am closing out this PR as I have split it up in https://github.com/dimagi/commcare-hq/pull/34663 and https://github.com/dimagi/commcare-hq/pull/34673
